File: targets/node/koa/routes/middleware.js
// middleware: per-layer dispatch cost at 0, 4 and 16 no-op layers.
//
// @koa/router takes middleware inline on the route, which is the scoping the family needs.
// Each layer awaits next() and does nothing else.
import * as d from "../../_shared/domain.js";

// rb:wiring middleware.*
const noop = async (_ctx, next) => { await next(); };

// rb:wiring middleware.*
const layers = (n) => Array.from({ length: n }, () => noop);

// rb:wiring middleware.*
const small = (ctx) => { ctx.body = d.payload("small"); };

export default function middleware(router) {
  router.get("/middleware/none", small);

  router.get("/middleware/four", ...layers(4), small);

  router.get("/middleware/sixteen", ...layers(16), small);
}

File: targets/java/vertx/src/main/java/rb/vertx/routes/Middleware.java
package rb.vertx.routes;

import io.vertx.ext.web.Router;
import rb.domain.Domain;
import rb.vertx.Reply;

/**
 * middleware: per-layer dispatch cost at 0, 4 and 16 no-op layers.
 *
 * Vert.x chains handlers on a route, which is the scoping the family needs. Each layer
 * calls next and does nothing else.
 */
public final class Middleware {
  private Middleware() {}

  // rb:wiring middleware.*
  private static void layered(Router router, String path, int n) {
    var route = router.get(path);
    for (int i = 0; i < n; i++) {
      route.handler(ctx -> ctx.next());
    }
    route.handler(ctx -> Reply.json(ctx, 200, Domain.payload("small")));
  }

  public static void register(Router router) {
    router.get("/middleware/none")
          .handler(ctx -> Reply.json(ctx, 200, Domain.payload("small")));

    // rb:handler middleware.four
    // rb:wiring middleware.*
    layered(router, "/middleware/four", 4);

    // rb:handler middleware.sixteen
    // rb:wiring middleware.*
    layered(router, "/middleware/sixteen", 16);
  }
}
